File: services/api/flux_watch_api/schema/alerts.py
import enum
import uuid
from datetime import datetime

# from uuid import UUID
from sqlalchemy import ARRAY, UUID, DateTime, Enum, Text
from sqlalchemy.orm import Mapped, mapped_column

from flux_watch_api.schema.mixins.parent_mixin import ParentMixin
from flux_watch_api.schema.utils.base import Base


class AlertType(enum.Enum):
    ERROR_SPIKE = "ErrorSpike"


class AlertLevel(enum.Enum):
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"
    SEVERE = "SEVERE"


class AlertORM(Base, ParentMixin):
    __tablename__ = "alerts"

    alert_type: Mapped[AlertType] = mapped_column(Enum(AlertType), nullable=False)
    level: Mapped[AlertLevel] = mapped_column(Enum(AlertLevel), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    triggered_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    event_ids: Mapped[list[uuid.UUID]] = mapped_column(ARRAY(UUID), nullable=False)

File: services/analytics_engine/src/alerts/alerts.rs
use crate::events::events::Parent;
use time::UtcDateTime;
use uuid::Uuid;

#[derive(Debug, PartialEq, PartialOrd, Clone, Copy)]
pub enum AlertLevel {
    High,
    Critical,
    Severe,
}

#[derive(Debug)]
pub enum AlertType {
    ErrorSpike,
}

#[derive(Debug)]
pub struct Alert {
    pub alert_type: AlertType,
    pub level: AlertLevel,
    pub message: String,
    pub triggered_at: UtcDateTime,
    pub event_ids: Vec<Uuid>,
    pub parent: Parent,
}

impl Alert {
    pub fn build_alert(
        alert_type: AlertType,
        level: AlertLevel,
        message: String,
        now: UtcDateTime,
        event_ids: Vec<Uuid>,
        parent: &Parent,
    ) -> Self {
        Self {
            alert_type,
            level,
            message,
            triggered_at: now,
            event_ids,
            parent: parent.clone(),
        }
    }
}
